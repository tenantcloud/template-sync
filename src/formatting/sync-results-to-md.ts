import { Change } from "diff";
import {
  TEMPLATE_SYNC_LOCAL_CONFIG,
  TemplateSyncReturn,
} from "../template-sync";

export function syncResultsToMd(result: TemplateSyncReturn) {
  return `# ${TEMPLATE_SYNC_LOCAL_CONFIG} 

## Stopped the following files from syncing:

${result.localSkipFiles.reduce((s, file) => {
  return `${s}* ${file}\n`;
}, "")}

## Changed the following files from what they would have synced:

${Object.keys(result.localFileChanges).reduce((s, file) => {
  return `${s}${file}
\`\`\`diff
${result.localFileChanges[file].reduce((diffS, change) => {
  return `${diffS}\n${makeChangeIntoDiffLines(change)}`;
}, "")}
\`\`\``;
}, "")}
`;
}

function makeChangeIntoDiffLines(change: Change, nonchangeMax = 3) {
  const operator = change.added ? "+" : change.removed ? "-" : "";

  if (
    !operator &&
    nonchangeMax > 0 &&
    change.count &&
    change.count > nonchangeMax
  ) {
    const lines = change.value.split("\n");
    const partial = lines.slice(lines.length - nonchangeMax);
    return `...\n${partial.join("\n")}\n`;
  }

  return change.value.split("\n").reduce((s, line) => {
    return `${s}${operator}${line}\n`;
  }, "");
}
