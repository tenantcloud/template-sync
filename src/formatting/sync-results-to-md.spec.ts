import { syncResultsToMd } from "./sync-results-to-md";

describe("syncResultsToMd", () => {
  it("snapshots", () => {
    expect(
      syncResultsToMd({
        localSkipFiles: ["src/file1.ts"],
        localFileChanges: {
          "package.json": [
            {
              count: 8,
              value: "my thing",
              added: true,
            },
            {
              count: 10,
              value: "your thang",
              removed: true,
            },
          ],
        },
      }),
    ).toMatchSnapshot();
  });
});
