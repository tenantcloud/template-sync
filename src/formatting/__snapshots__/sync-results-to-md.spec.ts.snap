// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syncResultsToMd snapshots 1`] = `
"# templatesync.local 

## Stopped the following files from syncing:

* src/file1.ts


## Changed the following files from what they would have synced:

package.json
\`\`\`diff

+my thing

-your thang

\`\`\`
"
`;
